host/frontend/host-orchestrator/operation.go

```diff
@@ -20,83 +20,126 @@ import (
+	Error OperationResultError
 }
 
-type OperationErrorData struct {
+type OperationResultError struct {
 	ErrorMsg string
 }
 
-type OperationOKData struct{}
-
 type OperationManager interface {
-	NewOperation() (OperationData, error)
+	New() Operation
+
+	Get(string) (Operation, error)
 
-	GetOperation(string) (OperationData, bool)
+	Complete(string, OperationResult)
+
+	// TODO(b/231319087) This should handle timeout.
+	Wait(string)
+}
 
-	CompleteOperation(string, OperationResultData)
+type mapOMOperationEntry struct {
+	data  Operation
+	mutex sync.RWMutex
+	wait  sync.WaitGroup
 }
 
 type MapOM struct {
 	uuidFactory func() string
 	mutex       sync.RWMutex
-	operations  map[string]OperationData
+	operations  map[string]*mapOMOperationEntry
 }
 
 func NewMapOM(uuidFactory func() string) *MapOM {
 	return &MapOM{
 		uuidFactory: uuidFactory,
 		mutex:       sync.RWMutex{},
-		operations:  make(map[string]OperationData),
+		operations:  make(map[string]*mapOMOperationEntry),
 	}
 }
 
-func (m *MapOM) NewOperation() (OperationData, error) {
+const mapOMNewUUIDRetryLimit = 100
+
+func (m *MapOM) New() Operation {
 	m.mutex.Lock()
 	defer m.mutex.Unlock()
-	name := m.uuidFactory()
-	if _, ok := m.operations[name]; ok {
-		return OperationData{}, NewOperationError("new operation name already exists")
+	name, retryCount := m.uuidFactory(), 0
+	_, found := m.operations[name]
+	for found && retryCount < mapOMNewUUIDRetryLimit {
+		name, retryCount = m.uuidFactory(), retryCount+1
+		_, found = m.operations[name]
+	}
+	if retryCount == mapOMNewUUIDRetryLimit {
+		panic("uuid retry limit reached")
+	}
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	op := &mapOMOperationEntry{
+		data: Operation{
+			Name:   name,
+			Done:   false,
+			Result: OperationResult{},
+		},
+		mutex: sync.RWMutex{},
+		wait:  wg,
 	}
-	op := OperationData{name, false, OperationResultData{}}
 	m.operations[name] = op
-	return op, nil
+	return op.data
 }
 
-func (m *MapOM) GetOperation(name string) (OperationData, bool) {
-	m.mutex.RLock()
-	defer m.mutex.RUnlock()
-	op, ok := m.operations[name]
-	return op, ok
+func (m *MapOM) Get(name string) (op Operation, err error) {
+	entry, ok := m.getOperationEntry(name)
+	if !ok {
+		err = NotFoundOperationError("map key didn't exist")
+		return
+	}
+	entry.mutex.RLock()
+	defer entry.mutex.RUnlock()
+	op = entry.data
+	return
 }
 
-func (m *MapOM) CompleteOperation(name string, result OperationResultData) {
-	m.mutex.Lock()
-	defer m.mutex.Unlock()
-	op, ok := m.operations[name]
+func (m *MapOM) Complete(name string, result OperationResult) {
+	op, ok := m.getOperationEntry(name)
 	if !ok {
 		log.Println("complete operation error: attempting to complete an operation which does not exist")
 		return
 	}
-	op.Done = true
-	op.Result = result
-	m.operations[name] = op
+	op.mutex.Lock()
+	defer op.mutex.Unlock()
+	op.wait.Done()
+	op.data.Done = true
+	op.data.Result = result
+}
+
+func (m *MapOM) Wait(name string) {
```

Comment:
The interface would be more useful if the (new) operation data were to be returned when `Wait()` finishes. It seems to me in almost every case calling this function would be immediately followed with a call to `Get()` otherwise.